tidy3d/components/data/near2far_radiation_vectors.py

```diff
@@ -485,345 +492,169 @@ def integrate_for_one_theta(i_th: int):
         """
-        r = np.sqrt(x**2 + y**2 + z**2)
-        theta = np.arccos(z / r)
-        phi = np.arctan2(y, x)
-        return r, theta, phi
 
-    @staticmethod
-    def _sph_2_car(r, theta, phi):
-        """coordinates only
+        freqs = np.atleast_1d(self.frequencies)
+        ux = np.atleast_1d(monitor.ux)
+        uy = np.atleast_1d(monitor.uy)
 
-        Parameters
-        ----------
-        r : float
-            radius.
-        theta : float
-            polar angle (rad) downward from x=y=0 line.
-        phi : float
-            azimuthal (rad) angle from y=z=0 line.
+        # compute radiation vectors for the dataset associated with each monitor
+        N_theta = np.zeros((len(ux), len(uy), len(freqs)), dtype=complex)
+        N_phi = np.zeros_like(N_theta)
+        L_theta = np.zeros_like(N_theta)
+        L_phi = np.zeros_like(N_theta)
 
-        Returns
-        -------
-        tuple
-            x, y, and z in cartesian coordinates.
-        """
-        r_sin_theta = r * np.sin(theta)
-        x = r_sin_theta * np.cos(phi)
-        y = r_sin_theta * np.sin(phi)
-        z = r * np.cos(theta)
-        return x, y, z
+        # Zip together all combinations of observation points for better progress tracking
+        iter_coords = [([_ux, _uy], [i, j]) for i, _ux in enumerate(ux) for j, _uy in enumerate(uy)]
 
-    @staticmethod
-    def _sph_2_car_field(Ar, Atheta, Aphi, theta, phi):
-        """Convert vector field components in spherical coordinates to cartesian.
+        for (_ux, _uy), (i, j) in track(iter_coords, description="Computing radiation vectors"):
+            theta, phi = AbstractNear2FarData.kspace_2_sph(_ux, _uy, monitor.u_axis)
 
-        Parameters
-        ----------
-        Ar : float
-            radial component of vector A.
-        Atheta : float
-            polar angle component of vector A.
-        Aphi : float
-            azimuthal angle component of vector A.
-        theta : float
-            polar angle (rad) of location of A.
-        phi : float
-            azimuthal angle (rad) of location of A.
-
-        Returns
-        -------
-        tuple
-            x, y, and z components of A in cartesian coordinates.
-        """
-        sin_theta = np.sin(theta)
-        cos_theta = np.cos(theta)
-        sin_phi = np.sin(phi)
-        cos_phi = np.cos(phi)
-        Ax = Ar * sin_theta * cos_phi + Atheta * cos_theta * cos_phi - Aphi * sin_phi
-        Ay = Ar * sin_theta * sin_phi + Atheta * cos_theta * sin_phi + Aphi * cos_phi
-        Az = Ar * cos_theta - Atheta * sin_theta
-        return Ax, Ay, Az
+            for surface in self.surfaces:
+                for idx_f, frequency in enumerate(freqs):
+                    _N_th, _N_ph, _L_th, _L_ph = self._radiation_vectors_for_surface(
+                        frequency, theta, phi, surface, self.currents[surface.monitor.name]
+                    )
+                    N_theta[i, j, idx_f] += _N_th
```

Comment:
I think for now this is ok, but there seems to be a pattern in here where the same operations are copied and pasted over a set of variables (`N_theta`, `N_phi`, `L_theta, `L_phi` in this case).  There's nothing wrong with this, but it would probably help to move towards using the data structures and comprehensions (where applicable) to reduce code redundancy and make things a bit more general.  For example 
```py
rad_vector_names = ('Ntheta', 'Nphi', 'Ltheta', 'Lphi')
rad_vectors = {name: np.zeros((len(ux), len(uy), len(freqs)), dtype=complex) for name in rad_vector_names}
...
for ...
    rad_vectors_surface = self._radiation_vectors_for_surface(...)
    for rad_vector, rad_vector_surface in zip(rad_vectors, rad_vectors_surface):
        rad_vector[i, j, idx_f] += rad_vector_surface

fields = {name:  Near2FarKSpaceDataArray(rad_vector, coords=coords) for name, rad_vector in rad_vectors.items()}
return Near2FarKSpaceData(monitor=monitor, **fields)
```

It would be a lot of work to refactor this, but I think it would be good to keep this in mind in the future if realizing that a lot of lines are getting copied and pasted.  Just keeps things more compact and maintainable.